controller/src/main/java/io/pravega/controller/task/Stream/StreamTransactionMetadataTasks.java

```diff
@@ -199,71 +209,97 @@ public Void initializeStreamWriters(final String commitStreamName, final EventSt
-                () -> commitTxnBody(scope, stream, txId, context));
+    private CompletableFuture<Void> checkReady() {
+        if (!ready) {
+            return FutureHelpers.failedFuture(new IllegalStateException(getClass().getName() + " not yet ready"));
+        } else {
+            return CompletableFuture.completedFuture(null);
+        }
     }
 
-    private CompletableFuture<Pair<VersionedTransactionData, List<Segment>>> createTxnBody(final String scope, final String stream,
-                                                                      final long lease, final long maxExecutionPeriod,
-                                                                      final long scaleGracePeriod,
-                                                                      final OperationContext context) {
+    CompletableFuture<Pair<VersionedTransactionData, List<Segment>>> createTxnBody(final String scope,
+                                                                                   final String stream,
+                                                                                   final long lease,
+                                                                                   final long maxExecutionPeriod,
+                                                                                   final long scaleGracePeriod,
+                                                                                   final OperationContext ctx) {
         UUID txnId = UUID.randomUUID();
-        return streamMetadataStore.createTransaction(scope, stream, txnId, lease, maxExecutionPeriod, scaleGracePeriod, context, executor)
-                .thenCompose(txData ->
-                        streamMetadataStore.getActiveSegments(scope, stream, context, executor)
-                                .thenCompose(activeSegments ->
-                                        FutureHelpers.allOf(
-                                                activeSegments.stream()
-                                                        .parallel()
-                                                        .map(segment ->
-                                                                notifyTxCreation(scope,
-                                                                        stream,
-                                                                        segment.getNumber(),
-                                                                        txData.getId()))
-                                                        .collect(Collectors.toList()))
-                                                .thenApply(v -> new ImmutablePair<>(txData, activeSegments))));
+        TxnResource resource = new TxnResource(scope, stream, txnId);
+        CompletableFuture<Void> addIndex = streamMetadataStore.addTxnToIndex(hostId, resource, 0);
+        return addIndex.thenComposeAsync(x -> streamMetadataStore.createTransaction(scope, stream, txnId, lease,
```

Comment:
We have nested `thenComposeAsync` calls, which causes additional indentation in each call.